extensions/amp-access/0.1/amp-access-server.js

```diff
@@ -0,0 +1,169 @@
+ * limitations under the License.
+ */
+
+import {AccessClientAdapter} from './amp-access-client';
+import {isProxyOrigin, removeFragment} from '../../../src/url';
+import {dev} from '../../../src/log';
+import {timer} from '../../../src/timer';
+import {vsyncFor} from '../../../src/vsync';
+import {xhrFor} from '../../../src/xhr';
+
+/** @const {string} */
+const TAG = 'amp-access-server';
+
+/** @const {number} */
+const AUTHORIZATION_TIMEOUT = 3000;
+
+
+/** @implements {AccessTypeAdapterDef} */
+export class AccessServerAdapter {
+
+  /**
+   * @param {!Window} win
+   * @param {!JSONObject} configJson
+   * @param {!AccessTypeAdapterContextDef} context
+   */
+  constructor(win, configJson, context) {
+    /** @const {!Window} */
+    this.win = win;
+
+    /** @const @private {!AccessTypeAdapterContextDef} */
+    this.context_ = context;
+
+    /** @private @const */
+    this.clientAdapter_ = new AccessClientAdapter(win, configJson, context);
+
+    /** @const @private {!Xhr} */
+    this.xhr_ = xhrFor(win);
+
+    /** @const @private {!Timer} */
+    this.timer_ = timer;
+
+    /** @const @private {!Vsync} */
+    this.vsync_ = vsyncFor(win);
+
+    const stateElement = this.win.document.querySelector(
+        'meta[name="i-amp-access-state"]');
+
+    /** @private @const {?string} */
+    this.serverState_ = stateElement ?
+        stateElement.getAttribute('content') : null;
+
+    /** @private @const {boolean} */
+    this.isProxyOrigin_ = isProxyOrigin(this.win.location);
+
+    /** @private @const {string} */
+    this.serviceUrl_ = removeFragment(this.win.location.href);
+  }
+
+  /** @override */
+  getConfig() {
+    return {
+      'client': this.clientAdapter_.getConfig(),
+      'proxy': this.isProxyOrigin_,
+      'serverState': this.serverState_,
+    };
+  }
+
+  /** @override */
+  isAuthorizationEnabled() {
+    return true;
+  }
+
+  /** @override */
+  authorize() {
+    dev.fine(TAG, 'Start authorization with ',
+        this.isProxyOrigin_ ? 'proxy' : 'non-proxy',
+        this.serverState_,
+        this.clientAdapter_.getAuthorizationUrl());
+    if (!this.isProxyOrigin_ || !this.serverState_) {
+      dev.fine(TAG, 'Proceed via client protocol');
+      return this.clientAdapter_.authorize();
+    }
+
+    dev.fine(TAG, 'Proceed via server protocol');
+
+    const varsPromise = this.context_.collectUrlVars(
+        this.clientAdapter_.getAuthorizationUrl(),
+        /* useAuthData */ false);
+    return varsPromise.then(vars => {
+      const requestVars = {};
+      for (const k in vars) {
+        if (vars[k] != null) {
+          requestVars[k] = String(vars[k]);
+        }
+      }
+      const request = {
+        'state': this.serverState_,
+        'vars': requestVars,
+      };
+      dev.fine(TAG, 'Authorization request: ', this.serviceUrl_, request);
+      // Note that `application/x-www-form-urlencoded` is used to avoid
+      // CORS preflight request.
```

Comment:
For experimentation we may need to deploy server on different origin temporarily. So it might be CORS. However, we definitely don't need `credentials: include` - i'll remove.